opt_einsum/backends/array_api.py

```diff
@@ -0,0 +1,72 @@
+"""
+Required functions for optimized contractions of arrays using array API-compliant backends.
+"""
+import sys
+from typing import Callable
+from types import ModuleType
+
+import numpy as np
+
+from ..sharing import to_backend_cache_wrap
+
+
+def discover_array_apis():
+    """Discover array API backends."""
+    if sys.version_info >= (3, 8):
+        from importlib.metadata import entry_points
+
+        if sys.version_info >= (3, 10):
+            eps = entry_points(group="array_api")
+        else:
+            # Deprecated - will raise warning in Python versions >= 3.10
+            eps = entry_points().get("array_api", [])
+        return [ep.load() for ep in eps]
+    else:
+        # importlib.metadata was introduced in Python 3.8, so it isn't available here. Unable to discover any array APIs.
+        return []
+
+
+def make_to_array_function(array_api: ModuleType) -> Callable:
+    """Make a ``to_[array_api]`` function for the given array API."""
+
+    @to_backend_cache_wrap
+    def to_array(array):  # pragma: no cover
+        if isinstance(array, np.ndarray):
+            return array_api.asarray(array)
+        return array
+
+    return to_array
+
+
+def make_build_expression_function(array_api: ModuleType) -> Callable:
+    """Make a ``build_expression`` function for the given array API."""
+
+    def build_expression(_, expr):  # pragma: no cover
+        """Build an array API function based on ``arrays`` and ``expr``."""
+
+        def array_api_contract(*arrays):
+            return expr._contract([make_to_array_function(array_api)(x) for x in arrays], backend=array_api.__name__)
```

Comment:
Should `to_array = make_to_array_function(array_api)` maybe be assigned in the outer scope to avoid reconstructing it for every array? (and below)

Reply:
Yes, that's much better